contracts/src/CrossChainApplications/NativeTokenBridge/NativeTokenDestination.sol

```diff
@@ -0,0 +1,213 @@
+// solhint-disable-next-line
+import {IAllowList} from "@subnet-evm-contracts/interfaces/IAllowList.sol";
+
+// The address where the burned transaction fees are credited.
+// TODO implement mechanism to report burned tx fees to source chian.
+address constant BURNED_TX_FEES_ADDRESS = 0x0100000000000000000000000000000000000000;
+// Designated Blackhole Address. Tokens are sent here to be "burned" before sending an unlock
+// message to the source chain. Different from the burned tx fee address so they can be
+// tracked separately.
+address constant BLACKHOLE_ADDRESS = 0x0100000000000000000000000000000000000001;
+
+contract NativeTokenDestination is
+    ITeleporterReceiver,
+    INativeTokenDestination,
+    ReentrancyGuard
+{
+    INativeMinter private immutable _nativeMinter =
+        INativeMinter(0x0200000000000000000000000000000000000001);
+
+    uint256 public constant TRANSFER_NATIVE_TOKENS_REQUIRED_GAS = 100_000;
+    bytes32 public immutable sourceBlockchainID;
+    address public immutable nativeTokenSourceAddress;
+    // The first `initialReserveImbalance` tokens sent to this subnet will not be minted.
+    // `initialReserveImbalance` should be constructed to match the initial token supply of this subnet.
+    // This means tokens will not be minted until the source contact is collateralized.
+    uint256 public immutable initialReserveImbalance;
+    uint256 public currentReserveImbalance;
+    uint256 public totalMinted = 0;
+
+    // Used for sending and receiving Teleporter messages.
+    ITeleporterMessenger public immutable teleporterMessenger;
+
+    constructor(
+        address teleporterMessengerAddress,
+        bytes32 sourceBlockchainID_,
+        address nativeTokenSourceAddress_,
+        uint256 initialReserveImbalance_
+    ) {
+        require(
+            teleporterMessengerAddress != address(0),
+            "NativeTokenDestination: zero TeleporterMessenger address"
+        );
+        teleporterMessenger = ITeleporterMessenger(teleporterMessengerAddress);
+
+        require(
+            sourceBlockchainID_ != bytes32(0),
+            "NativeTokenDestination: zero source chain ID"
+        );
+        require(
+            sourceBlockchainID_ !=
+                IWarpMessenger(0x0200000000000000000000000000000000000005)
+                    .getBlockchainID(),
+            "NativeTokenDestination: cannot bridge with same blockchain"
+        );
+        sourceBlockchainID = sourceBlockchainID_;
+
+        require(
+            nativeTokenSourceAddress_ != address(0),
+            "NativeTokenDestination: zero source contract address"
+        );
+        nativeTokenSourceAddress = nativeTokenSourceAddress_;
+
+        initialReserveImbalance = initialReserveImbalance_;
+        currentReserveImbalance = initialReserveImbalance_;
+    }
+
+    /**
+     * @dev See {ITeleporterReceiver-receiveTeleporterMessage}.
+     *
+     * Receives a Teleporter message and routes to the appropriate internal function call.
```

Comment:
This comment is not entirely accurate, since most of the logic is handled in this method and there isn't any routing.